README.md
# esxi-vm-deploy
Offering a golang program to automate vm installation on a esxi host.
The program will be asking for some informations, templating a .vmk file, creating a virtual disk and installing operating system via embedded BOOTP python server.
Before starting the automatic deployment of virtual machines you will need to tune up your system in order to be able to run program.
To run this program you will need:
* go > 1.13
* ansible >= 2.7

## Automatic System Setup
If you are running ***Debian***, in order to setup the system and be able to running golang program you can run the script below (needs sudo privilege on host) for automatic setup or you can follow the step-by-step guide in "Manual Setup" section
```
curl https://raw.githubusercontent.com/lucabodd/esxi-vm-deploy/master/setup/ESXi-vm-deploy-install.sh | bash
```

## Manual System Setup
If automatic system setup fails, or you want to setup the system manually, you can setup the system following the steps below.
In order to setup the system and being able to running golang program you will need to follow the following steps:
install ansible via package manager (Debian):
```
sudo apt-get install ansible
```
Install golang from sources
```
wget https://dl.google.com/go/go1.14.4.linux-amd64.tar.gz
tar -xvf go1.14.4.linux-amd64.tar.gz
sudo mv go /usr/local
rm -rf go
```
Install esxi-vm-deploy
```
go get github.com/lucabodd/esxi-vm-deploy
go install github.com/lucabodd/esxi-vm-deploy
```
In some cases you might need to export this environment vars
```
export GOROOT=/usr/local/go
export GOPATH=$HOME/go
export PATH=$GOPATH/bin:$GOROOT/bin:$PATH
```

## Update
As of today updates will be released on github.com, in order to update the program you can run the following command
```
go get -u github.com/lucabodd/esxi-vm-deploy
```

## Usage
Usage is described running the program with the help flag.
Before running the program, please consider the following checklist:
* "helper" host and "vm-ip" must be in the same target subnet (unless you are able to broadcast UDP packets :67,:68 outside the helper's server)
* one helper can deploy one vm per time, if you want to deploy more than one VM at once, you'll need more helpers (or wait :'))

```
Usage: esxi-vm-deploy [OPTIONS]
  -cpu string
        Specify RAM size (default "2")
  -disk-size string
        Virtual machine Disk size (default "50")
  -esxi string
        ESXi Hypervisor
  -help
        prints this help message
  -helper string
        BOOTP server address, specified host will provide configurations to booting (PXE) virtual machine
  -ip string
        Virtual machine IP address
  -name string
        Specify virtual machine name
  -os string
        Specify virtual machine OS available: debian9-64, debian10-64, debian11-64 (default "debian11-64")
  -ram string
        Specify RAM size (default "16")
  -use-custom-mirror string
        Define a custom mirror where images will be retriven from. Mirror must be defined as full path, point to amd64 directori, which must contain initrd.gz file (eg: http://ftp.nl.debian.org/debian/dists/bookworm/main/installer-amd64/current/images/netboot/debian-installer or https://d-i.debian.org/daily-images/amd64/daily/netboot/debian-installer/amd64/ )
  -use-default-mirror
        Download debian images from mirror, default mirror used is http://ftp.nl.debian.org if you want to use an alternative mirror use --mirror flag
  -v    enable verbose mode
  -version
        Display current version of script
  -vi-preseed
        edit local preseed file (changes will be overwritten during updates)
One ore more required flag has not been prodided.
Note that using less flag than required could lead program into errors
Omit flags only if you are aware of what are you doin'
[EXAMPLES]
- Creation of machine with custom hardware
esxi-vm-deploy --esxi [ESXi host defined in ssh config] -H [unix host defined in ssh config]  --ip [ip of new machine] --name [name of new machine] --ram 8  --disk-size 16 --cpu 4
- Creation of machine with default values 3 CPU 50GB Disk 16GB RAM
esxi-vm-deploy --esxi [ESXi host defined in ssh config] -H [unix host defined in ssh config]  --ip [ip of new machine] --name [name of new machine]
- Creation of machine with default values 3 CPU 50GB Disk 16GB RAM using the default mirror (http://ftp.nl.debian.org/)
esxi-vm-deploy --esxi [ESXi host defined in ssh config] -H [unix host defined in ssh config]  --ip [ip of new machine] --name [name of new machine] --use-default-mirror
- Creation of machine with default values 3 CPU 50GB Disk 16GB RAM using a custom mirror
esxi-vm-deploy --esxi [ESXi host defined in ssh config] -H [unix host defined in ssh config]  --ip [ip of new machine] --name [name of new machine] --os debian9-64 --use-custom-mirror http://ftp.debian.org/debian/dists/stretch/main/installer-amd64/current/images/netboot/
```

## Maintainance
To maintain this tool you need to periodically update debian netboot images in order to allow the the kernel version of the installer to match the version available in the debian archives.
Download the new installer version from debian repos.
```
wget --recursive --no-parent -nH --cut-dirs=8 --reject html,tmp http://ftp.nl.debian.org/debian/dists/stretch/main/installer-amd64/current/images/netboot/debian-installer/
```
and put the debian-installer directory inside ```esxi-vm-deploy/roles/vm-deploy/files/linux/images/debian[version]-64```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [2.0.0] - 2021-09-30
### Changed
- Renamed virtual machine flags with shorter names, removing 'vm-' prefix
- Installer images location/structure
- Unified preseed.cfg file
- Unified pxe configurations
- Prebuilt ldlinux.c32
- Python libs netifaces and numpy are now uploaded from the local package
- Upload module use less network bandwidth discriminating between installer versions

### Added
- Debian installers are now downloadable from online sources using a default mirror (ftp.debian.nl) using --use-default-mirror flag
- Debian installers are now downloadable from online sources using a custor mirror given from command line using --use-custom-mirror flag (refer to --help for further details)
- --vi-preseed allow users to edit package's preseed file (local changes will be overwritten during updates)


## [1.2.10] - 2021-08-14
### Changed
- Updating debian10-64 images in order to match kernel version with repos
### Added
- debian11-64 images

## [1.2.10] - 2021-07-14
### Changed
- Changing default keymap to 'en' in preseed.cfg

## [1.2.9] - 2021-06-14
### Changed
- Updating debian10-64 images in order to match kernel version with repos

## [1.2.8] - 2021-06-11
### Changed
- Setting LVM as default volume manager for both debian10-64 and debian9-64

## [1.2.7] - 2020-10-20
### Changed
- Updating debian10-64 images in order to match kernel version with repos

## [1.2.6] - 2020-10-20
### Changed
- changing repo name and shortening sleep time

## [1.2.5] - 2020-10-20
### Changed
- Updating debian10-64 images in order to match kernel version with repos

## [1.2.1] - 2020-01-29
### Added
- verbose mode

## [1.1.0] - 2019-10-19
### Added
- ESXi disk space check before create .vmdk disk

## [1.0.0] - 2019-06-20
### Added
- CHANGELOG.md

### Changed
- go-ansible lib has been changed from "github.com/apenella/go-ansible" to "github.com/lucabodd/go-ansible" (version hop, not backward compatible)


roles/files/linux/pybootd/pybootd/util.py
# Copyright (c) 2010-2019 Emmanuel Blot <emmanuel.blot@free.fr>
# Copyright (c) 2010-2011 Neotion
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

from configparser import ConfigParser, InterpolationSyntaxError
from logging import (DEBUG, INFO, ERROR, CRITICAL, WARNING,
                     Formatter, FileHandler, StreamHandler, getLogger)
from logging.handlers import (BufferingHandler, NTEventLogHandler,
                              SysLogHandler)
from re import match
from socket import inet_aton, inet_ntoa
from subprocess import run
from struct import pack as spack, unpack as sunpack
from sys import platform, stderr

try:
    import netifaces as nif
except ImportError:
    if platform == 'darwin':
        raise ImportError('netifaces package is not installed')
    nif = None


# String values evaluated as true boolean values
TRUE_BOOLEANS = ['on', 'high', 'true', 'enable', 'enabled', 'yes',  '1']
# String values evaluated as false boolean values
FALSE_BOOLEANS = ['off', 'low', 'false', 'disable', 'disabled', 'no', '0']
# ASCII or '.' filter
ASCIIFILTER = bytearray((''.join([(
    (len(repr(chr(_x))) == 3) or (_x == 0x5c)) and chr(_x) or '.'
    for _x in range(128)]) + '.' * 128).encode('ascii'))


def to_int(value):
    """Parse a value and convert it into an integer value if possible.

       Input value may be:
       - a string with an integer coded as a decimal value
       - a string with an integer coded as a hexadecimal value
       - a integral value
       - a integral value with a unit specifier (kilo or mega)
    """
    if not value:
        return 0
    if isinstance(value, int):
        return int(value)
    mo = match(r'^\s*(\d+)\s*(?:([KMkm]i?)?B?)?\s*$', value)
    if mo:
        mult = {'K': (1000),
                'KI': (1 << 10),
                'M': (1000 * 1000),
                'MI': (1 << 20)}
        value = int(mo.group(1))
        if mo.group(2):
            value *= mult[mo.group(2).upper()]
        return value
    return int(value.strip(), value.startswith('0x') and 16 or 10)


def to_bool(value, permissive=True, allow_int=False):
    """Parse a string and convert it into a boolean value if possible.

       :param value: the value to parse and convert
       :param permissive: default to the False value if parsing fails
       :param allow_int: allow an integral type as the input value

       Input value may be:
       - a string with an integer value, if `allow_int` is enabled
       - a boolean value
       - a string with a common boolean definition
    """
    if value is None:
        return False
    if isinstance(value, bool):
        return value
    if isinstance(value, int):
        if allow_int:
            return bool(value)
        else:
            if permissive:
                return False
            raise ValueError("Invalid boolean value: '%d'", value)
    if value.lower() in TRUE_BOOLEANS:
        return True
    if permissive or (value.lower() in FALSE_BOOLEANS):
        return False
    raise ValueError('"Invalid boolean value: "%s"' % value)


def hexline(data, sep=' '):
    """Convert a binary buffer into a hexadecimal representation

       Return a string with hexadecimal values and ASCII representation
       of the buffer data
    """
    try:
        if isinstance(data, bytes):
            src = bytearray(data)
        elif isinstance(data, bytearray):
            src = data
        elif isinstance(data, str):
            src = data.encode()
        else:
            # data may be a list/tuple
            src = bytearray(b''.join(data))
    except Exception:
        raise TypeError("Unsupported data type '%s'" % type(data))

    hexa = sep.join(["%02x" % x for x in src])
    printable = src.translate(ASCIIFILTER).decode('ascii')
    return "(%d) %s : %s" % (len(data), hexa, printable)


def logger_factory(logtype='syslog', logfile=None, level='WARNING',
                   logid='PXEd', format=None):
    # this code has been copied from Trac (MIT modified license)
    logger = getLogger(logid)
    logtype = logtype.lower()
    if logtype == 'file':
        hdlr = FileHandler(logfile)
    elif logtype in ('winlog', 'eventlog', 'nteventlog'):
        # Requires win32 extensions
        hdlr = NTEventLogHandler(logid, logtype='Application')
    elif logtype in ('syslog', 'unix'):
        hdlr = SysLogHandler('/dev/log')
    elif logtype in ('stderr'):
        hdlr = StreamHandler(stderr)
    else:
        hdlr = BufferingHandler(0)

    if not format:
        format = 'PXEd[%(module)s] %(levelname)s: %(message)s'
        if logtype in ('file', 'stderr'):
            format = '%(asctime)s ' + format
    datefmt = ''
    if logtype == 'stderr':
        datefmt = '%X'
    level = level.upper()
    if level in ('DEBUG', 'ALL'):
        logger.setLevel(DEBUG)
    elif level == 'INFO':
        logger.setLevel(INFO)
    elif level == 'ERROR':
        logger.setLevel(ERROR)
    elif level == 'CRITICAL':
        logger.setLevel(CRITICAL)
    else:
        logger.setLevel(WARNING)
    formatter = Formatter(format, datefmt)
    hdlr.setFormatter(formatter)
    logger.addHandler(hdlr)
    return logger


def iptoint(ipstr):
    return sunpack('!I', inet_aton(ipstr))[0]


def inttoip(ipval):
    return inet_ntoa(spack('!I', ipval))


def _netifaces_get_iface_config(address):
    pool = iptoint(address)
    for iface in nif.interfaces():
        ifinfo = nif.ifaddresses(iface)
        if nif.AF_INET not in ifinfo:
            continue
        for inetinfo in nif.ifaddresses(iface)[nif.AF_INET]:
            addr_s = inetinfo.get('addr')
            netmask_s = inetinfo.get('netmask')
            if addr_s is None or netmask_s is None:
                continue

            addr = iptoint(addr_s)
            mask = iptoint(netmask_s)
            ip = addr & mask
            ip_client = pool & mask
            delta = ip ^ ip_client
            if not delta:
                config = {'ifname': iface,
                          'server': inttoip(addr),
                          'net': inttoip(ip),
                          'mask': inttoip(mask)}
                return config
    return None


def _iproute_get_iface_config(address):
    pool = iptoint(address)
    iplines = (line.strip()
            for line in run("ip address show").stdout.split('\n'))
    iface = None
    for l in iplines:
        items = l.split()
        if not items:
            continue
        if items[0].endswith(':'):
            iface = items[1][:-1]
        elif items[0] == 'inet':
            saddr, smasklen = items[1].split('/', 1)
            addr = iptoint(saddr)
            masklen = int(smasklen)
            mask = ((1 << masklen) - 1) << (32 - masklen)
            ip = addr & mask
            ip_client = pool & mask
            delta = ip ^ ip_client
            if not delta:
                return {'ifname': iface,
                        'server': inttoip(addr),
                        'net': inttoip(ip),
                        'mask': inttoip(mask)}
    return None


def get_iface_config(address):
    if not address:
        return None
    #nifcfg = _netifaces_get_iface_config if nif else _iproute_get_iface_config
    nifcfg = _netifaces_get_iface_config
    return nifcfg(address)


def is_quoted(str_):
    """Tells whether a string is enclosed in simple- or double- quoted
       markers"""
    str_ = str_.strip()
    return (str_.startswith('"') and str_.endswith('"')) or \
           (str_.startswith("'") and str_.endswith("'"))


class EasyConfigParser(ConfigParser):
    """ConfigParser extension to support default config values and do not
       mess with multi-line option strings"""

    INDENT_SIZE = 8

    InterpolationSyntaxError = InterpolationSyntaxError

    def get(self, section, option, default=None, raw=True, vars=None,
            fallback=None):
        """Return the section:option value if it exists, or the default value
           if either the section or the option is missing"""
        if not self.has_section(section):
            return default
        if not self.has_option(section, option):
            return default
        return ConfigParser.get(self, section, option, raw=raw, vars=vars,
                                fallback=fallback)

    def write(self, filep):
        """Write an .ini-format representation of the configuration state,
           with automatic line wrapping, using improved multi-line
           representation.
        """
        for section in self._sections:
            filep.write("[%s]\n" % section)
            for (key, value) in self._sections[section].items():
                if key != "__name__":
                    filep.write("%s = %s\n" %
                                (key, str(value).replace('\n', '\n' +
                                 ' ' * self.INDENT_SIZE)))
            filep.write("\n")

    def _interpolate(self, section, option, rawval, vars):
        # special overloading of SafeConfigParser._interpolate:
        # do not attempt to interpolate if the string is (double-)quoted
        if is_quoted(rawval):
            return rawval
        # cannot use 'super' here as ConfigParser is outdated
        return ConfigParser._interpolate(self, section, option, rawval, vars)


esxi-vm-deploy.go
package main

import (
	"flag"
	"fmt"
	ansibler "github.com/lucabodd/go-ansible"
	"github.com/schollz/progressbar"
	"github.com/tidwall/gjson"
	"go/build"
	"log"
	"os"
	"strings"
	"strconv"
	"time"
	"os/exec"
)

func main() {
	//set ansible env vars
	os.Setenv("ANSIBLE_STDOUT_CALLBACK", "json")
	os.Setenv("ANSIBLE_HOST_KEY_CHECKING", "False")
	current_version := "2.0.0"

	//vars
	var esxi_host string
	var vm_name string
	var vm_os string
	var vm_net string
	var vm_datastore string
	var vm_cpu string
	var vm_ram string
	var vm_ipv4 string
	var vm_disk_size string
	var helper_host string
	var use_custom_mirror string
	var use_default_mirror bool
	var help bool
	var verbose bool
	var version bool
	var vi_preseed bool

	//Flag parsing
	flag.StringVar(&esxi_host, "esxi", "", "ESXi Hypervisor")
	flag.StringVar(&vm_name, "name", "", "Specify virtual machine name")
	flag.StringVar(&vm_os, "os", "debian11-64", "Specify virtual machine OS available: debian9-64, debian10-64, debian11-64")
	flag.StringVar(&vm_ram, "ram", "16", "Specify RAM size")
	flag.StringVar(&vm_cpu, "cpu", "2", "Specify RAM size")
	flag.StringVar(&vm_ipv4, "ip", "", "Virtual machine IP address")
	flag.StringVar(&vm_disk_size, "disk-size", "50", "Virtual machine Disk size")
	flag.StringVar(&helper_host, "helper", "", "BOOTP server address, specified host will provide configurations to booting (PXE) virtual machine")
	flag.BoolVar(&use_default_mirror, "use-default-mirror", false, "Download debian images from mirror, default mirror used is http://ftp.nl.debian.org if you want to use an alternative mirror use --mirror flag")
	flag.StringVar(&use_custom_mirror, "use-custom-mirror", "", "Define a custom mirror where images will be retriven from. Mirror must be defined as full path, point to amd64 directori, which must contain initrd.gz file (eg: http://ftp.nl.debian.org/debian/dists/bookworm/main/installer-amd64/current/images/netboot/debian-installer or https://d-i.debian.org/daily-images/amd64/daily/netboot/debian-installer/amd64/ )")
	flag.BoolVar(&version, "version", false, "Display current version of script")
	flag.BoolVar(&help, "help", false, "prints this help message")
	flag.BoolVar(&verbose, "v", false, "enable verbose mode")
	flag.BoolVar(&vi_preseed, "vi-preseed", false, "edit local preseed file (changes will be overwritten during updates)")
	flag.Parse()

	// retrive bin directory
	gopath := os.Getenv("GOPATH")
	if gopath == "" {
		gopath = build.Default.GOPATH
	}
	datadir := gopath+"/src/github.com/lucabodd/esxi-vm-deploy"


	if version {
		fmt.Println("esxi-vm-deploy version: ", current_version)
		fmt.Println("see CHANGELOG.md for latest version changes\ncopy available under "+datadir+"/CHANGELOG.md\nor at https://github.com/lucabodd/esxi-vm-deploy/blob/master/CHANGELOG.md")
		kill("")
	}
	if vi_preseed {
		cmd := exec.Command("vim", datadir+"/roles/vm-deploy/files/linux/preseed.cfgs/preseed.cfg")
    cmd.Stdin = os.Stdin
    cmd.Stdout = os.Stdout
    err := cmd.Run()
		check(err)
		kill("[+] Saved")
	}
	if use_default_mirror && use_custom_mirror!="" {
		fmt.Println("[*] Unable to determine the image source to use. User selected both --use-custom-mirror and --use-default-mirror. Using custom mirror")
		fmt.Println("[*] Warning: flags are misconfigured.")
	}
	if esxi_host == "" || vm_name == "" || vm_ipv4 == "" || helper_host == "" || help {
		fmt.Println("Usage: esxi-vm-deploy [OPTIONS]")
		flag.PrintDefaults()
		fmt.Println("One ore more required flag has not been prodided.")
		fmt.Println("Note that using less flag than required could lead program into errors \nOmit flags only if you are aware of what are you doin'")
		fmt.Println("[EXAMPLES]")
		fmt.Println("- Creation of machine with custom hardware")
		fmt.Println("esxi-vm-deploy --esxi [ESXi host defined in ssh config] -H [unix host defined in ssh config]  --ip [ip of new machine] --name [name of new machine] --ram 8  --disk-size 16 --cpu 4")
		fmt.Println("- Creation of machine with default values 3 CPU 50GB Disk 16GB RAM")
		fmt.Println("esxi-vm-deploy --esxi [ESXi host defined in ssh config] -H [unix host defined in ssh config]  --ip [ip of new machine] --name [name of new machine]")
		fmt.Println("- Creation of machine with default values 3 CPU 50GB Disk 16GB RAM using the default mirror (http://ftp.nl.debian.org/)")
		fmt.Println("esxi-vm-deploy --esxi [ESXi host defined in ssh config] -H [unix host defined in ssh config]  --ip [ip of new machine] --name [name of new machine] --use-default-mirror")
		fmt.Println("- Creation of machine with default values 3 CPU 50GB Disk 16GB RAM using a custom mirror")
		fmt.Println("esxi-vm-deploy --esxi [ESXi host defined in ssh config] -H [unix host defined in ssh config]  --ip [ip of new machine] --name [name of new machine] --os debian9-64 --use-custom-mirror http://ftp.debian.org/debian/dists/stretch/main/installer-amd64/current/images/netboot/")
		kill("")
	}
	//end of flag parsing

	/*
		############################################################################
		#							VARS COLLECTION 							   #
		############################################################################
	*/
	log.Println("[*] Environment Prechecks...")

	playbook := &ansibler.PlaybookCmd{
		Playbook:          datadir+"/playbooks/esxi-check-duplicate.yml",
		ConnectionOptions: &ansibler.PlaybookConnectionOptions{},
		Options:           &ansibler.PlaybookOptions{
			Inventory: esxi_host + ",",
			ExtraVars: map[string]interface{}{
				"vm_name": vm_name,
			},
		},
	}
	res, err := playbook.Run()
	check(err)
	err = res.PlaybookResultsChecks()
	check(err)
	verboseOut(res.RawStdout, verbose)
	duplicate_stdout := gjson.Get(res.RawStdout, "plays.0.tasks.1.hosts.*.stdout")
	duplicate := duplicate_stdout.Int()
	if duplicate > 0 {
		kill("[-] ERROR: A virtual machine with the same name already exis")
	}

	fmt.Println("[+] System checks passed ... starting")

	fmt.Println("[*] Gathering ESXi host info")
	playbook = &ansibler.PlaybookCmd{
		Playbook:          datadir+"/playbooks/esxi-gather-info.yml",
		ConnectionOptions: &ansibler.PlaybookConnectionOptions{},
		Options:           &ansibler.PlaybookOptions{
			Inventory: esxi_host + ",",
		},
	}
	res = &ansibler.PlaybookResults{}
	res,err = playbook.Run()
	check(err)
	err = res.PlaybookResultsChecks()
	check(err)
	verboseOut(res.RawStdout, verbose)
	esxi_vmnet := gjson.Get(res.RawStdout, "plays.0.tasks.0.hosts.*.stdout_lines")
	vmnets := esxi_vmnet.Array()
	esxi_datastores := gjson.Get(res.RawStdout, "plays.0.tasks.1.hosts.*.stdout_lines")
	datastores := esxi_datastores.Array()

	//VMNET selection
	if len(vmnets) == 1 {
		vm_net = vmnets[0].Str
		log.Println("[+] Only 1 available networks found, selecting " + vm_net)
	} else {
		log.Println("[*] Found more than one virtual network, choose where you want to deploy VM")
		fmt.Println("Available networks:")
		for index, element := range vmnets {
			fmt.Println(index, "-", element)
		}
		fmt.Print("Insert ID: ")
		var id int
		_, err := fmt.Scanf("%d", &id)
		check(err)
		vm_net = vmnets[id].Str

	}
	//DATASTORE selection
	if len(datastores) == 1 {
		vm_datastore = datastores[0].Str
		log.Println("[+] Only 1 available datastore found, selecting " + vm_datastore)
	} else {
		log.Println("[*] Found more than one datastore, choose where you want to deploy VM")
		fmt.Println("Available datastores:")
		for index, element := range datastores {
			fmt.Println(index, "-", element)
		}
		fmt.Print("Insert ID: ")
		var id int
		_, err := fmt.Scanf("%d", &id)
		check(err)
		vm_datastore = datastores[id].Str
	}

	//check datastore available space
	playbook = &ansibler.PlaybookCmd{
		Playbook:          datadir+"/playbooks/esxi-check-datastore.yml",
		ConnectionOptions: &ansibler.PlaybookConnectionOptions{},
		Options:           &ansibler.PlaybookOptions{
			Inventory: esxi_host + ",",
			ExtraVars: map[string]interface{}{
				"datastore": vm_datastore,
			},
		},
	}
	res = &ansibler.PlaybookResults{}
	res,err = playbook.Run()
	check(err)
	err = res.PlaybookResultsChecks()
	check(err)
	verboseOut(res.RawStdout, verbose)
	esxi_available_space := gjson.Get(res.RawStdout, "plays.0.tasks.0.hosts.*.stdout").Str

	//parse result and convert TB to GB
	var esxi_available_space_qty float64
	if strings.Contains(esxi_available_space, "T"){
		esxi_available_space = strings.Trim(esxi_available_space, "T")
		esxi_available_space_qty , err = strconv.ParseFloat(esxi_available_space, 32)
		check(err)
    	esxi_available_space_qty = esxi_available_space_qty * 1024
	} else if strings.Contains(esxi_available_space, "G"){
		esxi_available_space = strings.Trim(esxi_available_space, "G")
		esxi_available_space_qty , err = strconv.ParseFloat(esxi_available_space, 32)
		check(err)
	}

	vm_disk_size_qty , err := strconv.ParseFloat(vm_disk_size, 32)
	check(err)
	if vm_disk_size_qty > esxi_available_space_qty {
		fmt.Println("[-]ERROR: ESXi host has only %.2fG of free memory and cannot suite vm size", esxi_available_space_qty)
		kill("")
	}
	/*
		############################################################################
		#							VARS COLLECTION END 						   #
		############################################################################
	*/
	/*
		############################################################################
		#							VMX DEPLOYMENT  							   #
		############################################################################
	*/
	log.Println("[+] Requirements analysis passed, deploying .vmx and allocating disk space (thick)")
	log.Println("[*] Deploying .vmx file")
	//vmware still not supporting debian 11 this is a patch and shall be removed in the future
	vm_os_temp := vm_os
	if vm_os == "debian11-64" {
		vm_os = "debian10-64"
	}
	playbook = &ansibler.PlaybookCmd{
		Playbook:          datadir+"/playbooks/esxi-deploy-vmx.yml",
		ConnectionOptions: &ansibler.PlaybookConnectionOptions{},
		Options:           &ansibler.PlaybookOptions{
			Inventory: esxi_host + ",",
			ExtraVars: map[string]interface{}{
				"vm_name":      vm_name,
				"vm_os":        vm_os,
				"vm_cpu":       vm_cpu,
				"vm_disk_size": vm_disk_size,
				"vm_ram":       vm_ram,
				"vm_net":       vm_net,
				"vm_datastore": vm_datastore,
			},
		},
	}
	res = &ansibler.PlaybookResults{}
	res, err = playbook.Run()
	check(err)
	err = res.PlaybookResultsChecks()
	check(err)
	verboseOut(res.RawStdout, verbose)
	//vmware still not supporting debian 11 this is a patch and shall be removed in the future
	vm_os = vm_os_temp
	log.Println("[+] Virtual machine created. disk initialization completed")
	log.Println("[*] Retrieveing VM mac address")
	log.Println("[*] Retrieveing Assigned VM Hypervisor ID")
	mac_stdout := gjson.Get(res.RawStdout, "plays.0.tasks.4.hosts.*.stdout")
	vm_mac_addr := mac_stdout.String()
	id_stdout := gjson.Get(res.RawStdout, "plays.0.tasks.6.hosts.*.stdout")
	vm_id := id_stdout.String()
	log.Println("[+] Got physical address: " + vm_mac_addr)
	log.Println("[+] Got VMID: " + vm_id)
	/*
		############################################################################
		#							VMX DEPLOYMENT - END						   #
		############################################################################
	*/
	/*
		############################################################################
		#							PYBOOTD  DEPLOYMENT 						   #
		############################################################################
	*/
	log.Println("[*] uploading BOOTP server")
	mirror:=""

	if use_default_mirror {
		switch vm_os {
		case "debian11-64":
			mirror="http://ftp.nl.debian.org/debian/dists/bullseye/main/installer-amd64/current/images/netboot/debian-installer/"
		case "debian10-64":
			mirror="http://ftp.nl.debian.org/debian/dists/buster/main/installer-amd64/current/images/netboot/debian-installer/"
		case "debian9-64":
			mirror="http://ftp.nl.debian.org/debian/dists/stretch/main/installer-amd64/current/images/netboot/debian-installer/"
		}
	}
	if use_custom_mirror!="" {
		mirror=use_custom_mirror
	}

	//use bool var in order to track in ansible wether to use a mirror or upload images
	use_mirror:=false
	if mirror == "" {
		use_mirror=false
	} else {
		use_mirror=true
	}

	playbook = &ansibler.PlaybookCmd{
		Playbook:          datadir+"/playbooks/bootp-server-deploy.yml",
		ConnectionOptions: &ansibler.PlaybookConnectionOptions{},
		Options:           &ansibler.PlaybookOptions{
			Inventory: helper_host + ",",
			ExtraVars: map[string]interface{}{
				"vm_ipv4":     vm_ipv4,
				"vm_mac_addr": vm_mac_addr,
				"vm_name":     vm_name,
				"vm_os":        vm_os,
				"use_mirror":		use_mirror,
				"mirror":		mirror,
			},
		},
	}
	res = &ansibler.PlaybookResults{}
	res, err = playbook.Run()
	verboseOut(res.RawStdout, verbose)
	check(err)
	err = res.PlaybookResultsChecks()
	check(err)
	log.Println("[+] BOOTP server running")
	/*
		############################################################################
		#							PYBOOTD  DEPLOYMENT END						   #
		############################################################################
	*/
	/*
		############################################################################
		#							VM POWERUP     		 						   #
		############################################################################
	*/
	log.Println("[*] Powering up VM")
	playbook = &ansibler.PlaybookCmd{
		Playbook:          datadir+"/playbooks/esxi-vm-poweron.yml",
		ConnectionOptions: &ansibler.PlaybookConnectionOptions{},
		Options:           &ansibler.PlaybookOptions{
			Inventory: esxi_host + ",",
			ExtraVars: map[string]interface{}{
				"vm_id": vm_id,
			},
		},
	}
	res = &ansibler.PlaybookResults{}
	res, err = playbook.Run()
	check(err)
	err = res.PlaybookResultsChecks()
	check(err)
	verboseOut(res.RawStdout, verbose)
	log.Println("[+] VM " + vm_id + " powered on as " + vm_name)
	/*
		############################################################################
		#							VM POWERUP END 		 						   #
		############################################################################
	*/
	fmt.Println("\n")
	bar := progressbar.Default(100)
	for i := 0; i < 100; i++ {
		bar.Add(1)
		time.Sleep(900 * time.Millisecond)
	}
	fmt.Println("\n")
	/*
		############################################################################
		#							FILE CLEANUP 		 						   #
		############################################################################
	*/
	log.Println("[*] Running Helper deploy files cleanup")

	playbook = &ansibler.PlaybookCmd{
		Playbook:          datadir+"/playbooks/helper-depfiles-cleanup.yml",
		ConnectionOptions: &ansibler.PlaybookConnectionOptions{},
		Options:           &ansibler.PlaybookOptions{
			Inventory: helper_host + ",",
		},
	}
	res, err = playbook.Run()
    check(err)
	err = res.PlaybookResultsChecks()
    check(err)
	verboseOut(res.RawStdout, verbose)
	log.Println("[+] Cleanup Completed")
	/*
		############################################################################
		#							VM CLEANUP END 		 						   #
		############################################################################
	*/
}

func check(e error) {
	if e != nil {
		fmt.Println(e)
		panic(e)
	}
}
func kill(reason string) {
	fmt.Println(reason)
	os.Exit(1)
}

func verboseOut(message string, verbose bool){
	if verbose {
		fmt.Println(message)
	}
}
